File: scraper/src/common/lastfm/acquire.ts
import axios from 'axios';
import _ from 'lodash';
import qs from 'qs';

import { LASTFM_API_ERRORS, MAX_RETRIES } from '../constants.js';
import formatError from '../format-error.js';
import logger from '../logger.js';
import sleep from '../sleep.js';

import { DEFAULT_PARAMS } from './api-constants.js';
import type { Parameters, Payload } from './api-types.js';

const { isEmpty } = _;

const API_DELAY_MS = 1000;

const RESPONSE_TIMEOUT = 2000;
const CONNECTION_TIMEOUT = 5000;

let waiter = Promise.resolve();

export default async function acquire<T extends Payload>(
  parameters: Parameters,
  retry = 0,
): Promise<T | null> {
  await waiter;
  const url = `https://ws.audioscrobbler.com/2.0/?${qs.stringify({
    ...DEFAULT_PARAMS,
    ...parameters,
  })}`;

  logger.debug(url);
  try {
    const abortController = new AbortController();
    const connectionTimeout = setTimeout(() => {
      abortController.abort();
      logger.error('Aborted');
    }, CONNECTION_TIMEOUT);
    const response = await axios.get<T>(url, {
      signal: abortController.signal,
      timeout: RESPONSE_TIMEOUT,
    });
    clearTimeout(connectionTimeout);

    if (isEmpty(response?.data)) {
      throw new Error(response?.data?.message || 'Empty response');
    } else if (response.data.error || isEmpty(response.data)) {
      if (response.data.error === LASTFM_API_ERRORS.INVALID_PARAMETERS) {
        logger.error(response.data.message);
        return null;
      }
      throw new Error(response.data.message || 'Empty response');
    }
    waiter = sleep(API_DELAY_MS);
    return response.data;
  } catch (error) {
    logger.error(formatError(error));
    if (retry >= MAX_RETRIES) {
      throw error;
    }
    logger.warn(`retry #${retry + 1}`);
    // eslint-disable-next-line no-magic-numbers
    if (process.env.NODE_ENV !== 'test') await sleep(2 ** retry * 1000);
    return acquire<T>(parameters, retry + 1);
  }
}

File: scraper4/lastfm-api/lastfm_api/acquire.py
from os import environ
from time import sleep

from loguru import logger
from ratelimit import limits

CONNECTION_TIMEOUT = 5000
RESPONSE_TIMEOUT = 2000
MAX_RETRIES_NUMBER = 3
INVALID_PARAMETERS_ERROR = 6
DEFAULT_PARAMS = {
    "api_key": environ['LASTFM_API_KEY'],
    "autocorrect": '1',
    "format": 'json',
}


@limits(calls=1, period=1)
def acquire(params, retry_number=0):
    full_params = {}
    full_params.update(DEFAULT_PARAMS)
    full_params.update(params)
    url = f'https://ws.audioscrobbler.com/2.0/?{full_params}'
    logger.debug(url)
    try:
        response = requests.get(url, timeout=RESPONSE_TIMEOUT)
        if response.data is None:
            raise Exception(response.data['message'] or 'Empty response')
        elif response.data['error']:
            if response.data['error'] == INVALID_PARAMETERS_ERROR:
                logger.error(response.data['message'])
                return
            raise Exception(response.data['message'] or 'Empty response')
        return response.data
    except Exception as error:
        logger.error(error)
        if retry_number >= MAX_RETRIES_NUMBER:
            raise error
        logger.warning(f'retry #{retry_number + 1}')
        sleep(2 ** retry_number)
        return acquire(params, retry_number+1)
